.github/spam_cleaner.py

```diff
@@ -0,0 +1,443 @@
+
+
+def read_sheet(sheet, sheet_range):
+    """Read a sheet specified by the sheet_range."""
+    result_input = sheet.values().get(spreadsheetId=SPREADSHEET_ID, range=sheet_range).execute()
+    values_input = result_input.get("values", [])
+    df = pd.DataFrame(values_input[1:], columns=values_input[0])
+    return df.to_dict("split")["data"]
+
+
+def delete_verified_users(fetched_issues_data, verified_users_data):
+    """Delete all users who were verified from the passed dictionary."""
+    users_to_delete = []
+    for user, data in fetched_issues_data.items():
+        if (user, data["email"]) in verified_users_data:
+            users_to_delete.append(user)
+
+    for user in users_to_delete:
+        del fetched_issues_data[user]
+
+
+def update_sheet_and_send_notifications(sheet, fetched_issues_data, current_time):
+    """Update the sheet with newly added content, and newly added users and then notify them."""
+    users_to_verify_data = read_sheet(sheet, USERS_TO_VERIFY_SHEET_RANGE)
+    current_timestamp = datetime.datetime.timestamp(current_time)
+
+    rows_to_update, rows_to_append = separate_users_to_update_append_notify(
+        fetched_issues_data, users_to_verify_data, current_timestamp
+    )
+    request_data = prepare_update_request(
+        rows_to_update, rows_to_append, users_to_verify_data
+    )
+    send_batch_update_request(sheet, request_data)
+
+    users_to_notify = [email for _, email, _, _ in rows_to_append if email != 'Anonymous']
+    send_notifications(users_to_notify)
+
+
+def separate_users_to_update_append_notify(
+    fetched_issues_data, users_to_verify_data, current_timestamp
+):
+    """Prepare updated sheet's rows for existing users with recently added content and new users."""
+    rows_to_update = []
+    rows_to_append = []
+
+    # update already existing users' data and append new users' data
+    for fetched_user, fetched_data in fetched_issues_data.items():
+        updated = False
+        # a first data row starts at the position 2 because of the heading
+        last_row_number = 2
+        for user, email, data, _ in users_to_verify_data:
+            if (user, email) == (fetched_user, fetched_data["email"]):
+                data_list = json.loads(data)
+                data_list.extend(fetched_data["content"])
+                data_list_without_duplicates = remove_duplicates(data_list)
+                dumped_data = json.dumps(data_list_without_duplicates)
+                rows_to_update.append((last_row_number, dumped_data))
+                updated = True
+                break
+            last_row_number += 1
+
+        if not updated:
+            dumped_data = json.dumps(fetched_data["content"])
+            rows_to_append.append(
+                [fetched_user, fetched_data["email"], dumped_data, current_timestamp]
+            )
+
+    return rows_to_update, rows_to_append
+
+
+def remove_duplicates(data_list):
```

Comment:
Can you not just convert to a set and back?

Reply:
No, because a dictionary object is not hashable.

Reply:
No, because a dictionary object is not hashable. I am iterating over a list of dictionaries, e.g.: `[{"issue": "https://pulp.plan.io/issues/7188"}, {"journal": ["https://pulp.plan.io/issues/3215", 59794]}]`.